Add vivotek camera component

HarlemSquirrel opened the pull request:
## Description:

Add new integration for [Vivotek](https://www.vivotek.com/) cameras.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10189

## Example entry for `configuration.yaml` (if applicable):
```yaml
camera:
  - platform: vivotek
    name: Front door camera
    ip_address: 192.168.1.119
    username: !secret fd_camera_user
    password: !secret fd_camera_pwd
    verify_ssl: false
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

HarlemSquirrel committed:
Add vivotek camera component

HarlemSquirrel committed:
Update vivotek camera compontent

Use async request to enable/disable motion detection

HarlemSquirrel committed:
Update Vivotek camera

- Use HTTPS
- Use IP address for still and stream

HarlemSquirrel committed:
Update vivotek component

- Add brand and model properties
- Add state property
- Use attribute to save motion detection status

HarlemSquirrel committed:
Add vivotek camera to .coveragerc

homeassistant commented:
Hi @HarlemSquirrel, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#26071).

Once you do that we will be able to review and accept this pull request.

Thanks!

HarlemSquirrel committed:
Update vivotek camera

Fix lint errors

HarlemSquirrel committed:
Update vivotek camera

Remove unused method

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,257 @@
+    def get_param(self, param):
+        """Return the value of the provided key."""
+        try:
+            response = requests.get(
```
Please extract the device specific interface, I/O and data parsing, to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services

HarlemSquirrel commented:
OK, working on this https://github.com/HarlemSquirrel/python-vivotek

HarlemSquirrel committed:
Update Vivotek integration to use libpyvivotek

Use libpyvivotek instead of directly making HTTPS API calls.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+"""Support for Vivotek IP Cameras."""
+
+import logging
+import voluptuous as vol
```
Please add a blank line between standard library and 3rd party import groups.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+    def __init__(self, hass, device_info):
+        """Initialize a generic camera."""
+        from libpyvivotek import VivotekCamera
```
Please move 3rd party imports to the top of the module.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up a generic IP Camera."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+    """A Vivotek IP camera."""
+
+    def __init__(self, hass, device_info):
+        """Initialize a generic camera."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+class VivotekCamera(Camera):
+    """A Vivotek IP camera."""
+
+    def __init__(self, hass, device_info):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare replied:
Please rename `device_info` to `config`.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+        super().__init__()
+
+        self.hass = hass
```
Remove this.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+        self.hass = hass
+
+        self._name = device_info.get(CONF_NAME)
```
Please use `dict[key]` for required config keys and keys with default config schema values.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+    def camera_image(self):
+        """Return bytes of camera image."""
+        return run_coroutine_threadsafe(
```
```py
return self._cam.snapshot()
```

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+            self.async_camera_image(), self.hass.loop
+        ).result()
+
+    async def async_camera_image(self):
```
Remove this. The library doesn't support asyncio.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        """Return the camera motion detection status."""
+        return self._motion_detection_enabled
+
+    async def disable_motion_detection(self):
```
Don't implement the async api methods. Implement the sync methods.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        response = self._cam.set_param(self._event_0_key, 0)
+        self._motion_detection_enabled = int(response.replace("'", "")) == 1
+
+    async def enable_motion_detection(self):
```
See above.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+            pwd=password,
+        )
+
+        self._motion_detection_enabled = self.event_enabled(self._event_0_key)
```
We don't want side effects in init method. Please move this to `setup_platform` and pass in the result to the entity.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self._last_url = None
+        self._last_image = None
+
+        self._cam = VivotekCamera(
```
Instantiate the `VivotekCamera` in `setup_platform` and pass it in to the entity.

HarlemSquirrel replied:
I just realized the name of the class created in this module is the same as the imported from libpyvivotek so I need to rename it

MartinHjelmare replied:
We can rename the entity class.

MartinHjelmare replied:
Please instantiate the camera outside the entity and pass the camera in to the entity when creating it. The entity doesn't need to know about the credentials. Also build the stream source outside the entity and pass it in.

MartinHjelmare replied:
Bump.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+    def event_enabled(self, event_key):
+        """Return true if event for the provided key is enabled."""
+        response = self._cam.get_param(event_key)
+        return int(response.replace("'", "")) == 1
```
The library should parse the response from the camera. We shouldn't have to do that here in home assistant.

There should be separate methods for different features and they should return values that can be used without parsing, eg `True` or `False`.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        return self._cam.model_name
+
+    @property
+    def state(self):
```
We're not allowed to overwrite the state property. Please remove this.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self.hass = hass
+
+        self._name = device_info.get(CONF_NAME)
+        self._limit_refetch = device_info[CONF_LIMIT_REFETCH_TO_URL_CHANGE]
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self._name = device_info.get(CONF_NAME)
+        self._limit_refetch = device_info[CONF_LIMIT_REFETCH_TO_URL_CHANGE]
+        self._frame_interval = 1 / device_info[CONF_FRAMERATE]
+        self.content_type = device_info[CONF_CONTENT_TYPE]
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self._limit_refetch = device_info[CONF_LIMIT_REFETCH_TO_URL_CHANGE]
+        self._frame_interval = 1 / device_info[CONF_FRAMERATE]
+        self.content_type = device_info[CONF_CONTENT_TYPE]
+        self.verify_ssl = device_info[CONF_VERIFY_SSL]
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self._frame_interval = 1 / device_info[CONF_FRAMERATE]
+        self.content_type = device_info[CONF_CONTENT_TYPE]
+        self.verify_ssl = device_info[CONF_VERIFY_SSL]
+        self._event_i0_status = None
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+
+        self._supported_features = SUPPORT_STREAM if self._stream_source else 0
+
+        self._last_url = None
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+        self._supported_features = SUPPORT_STREAM if self._stream_source else 0
+
+        self._last_url = None
+        self._last_image = None
```
This isn't used.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,174 @@
+STATE_IDLE = "idle"
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```
Implement `def setup_platform` instead. The library doesn't support asyncio.

HarlemSquirrel committed:
Update Vivotek component

Address code review.

- Remove unused code
- Replace async methods with synchronous methods
- Update docstrings

HarlemSquirrel committed:
Linter fixes for Vivotek component

HarlemSquirrel committed:
Update Vivotek camera component

- Add SSL option
- Remove authentication options as only basic authentication
  is currently working

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,136 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_IP_ADDRESS): cv.string,
+        vol.Optional(CONF_STREAM_SOURCE, default=None): vol.Any(None, cv.string),
```
Don't use default `None`. Remove the default and use `dict.get` when accessing the config option.

MartinHjelmare replied:
What kind of string should the stream source option return? Does the user need to configure this?

HarlemSquirrel replied:
Ah yeah, good point. That can go.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,131 @@
+        self._motion_detection_enabled = False
+        self._event_0_key = DEFAULT_EVENT_0_KEY
+
+        username = config[CONF_USERNAME]
```
Username and password are optional and do not have defaults in the config schema. Then we need to use `dict.get`.

HarlemSquirrel replied:
Good point, username and password should be required.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,131 @@
+)
+
+STATE_DETECTING_MOTION = "detecting motion"
+STATE_IDLE = "idle"
```
We don't use these constants anymore.

HarlemSquirrel committed:
Update Vivotek camera component

- Make frame rate configurable
- Require username and password

HarlemSquirrel committed:
Remove unused constants in Vivotek component

HarlemSquirrel commented:
I believe all requested changes should be addressed now.

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,123 @@
+    def disable_motion_detection(self):
+        """Disable motion detection in camera."""
+        response = self._cam.set_param(self._event_0_key, 0)
+        self._motion_detection_enabled = int(response.replace("'", "")) == 1
```
Please handle the response parsing in the library.

HarlemSquirrel replied:
On it

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,123 @@
+)
+
+
+def setup_platform(hass, config, async_add_entities, discovery_info=None):
```
```suggestion
def setup_platform(hass, config, add_entities, discovery_info=None):
```

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,123 @@
+
+def setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up a Vivotek IP Camera."""
+    async_add_entities([VivotekCam(config)])
```
```suggestion
    add_entities([VivotekCam(config)])
```

HarlemSquirrel committed:
Update Vivotek camera integration

- Use libpyvivotek v0.2.1 with better response parsing
- Use add_entities instead of async_add_entities

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,123 @@
+        self._frame_interval = 1 / config[CONF_FRAMERATE]
+        self._name = config[CONF_NAME]
+        self._motion_detection_enabled = False
+        self._event_0_key = DEFAULT_EVENT_0_KEY
```
We don't need an instance attribute for the event key. We can use the constant `DEFAULT_EVENT_0_KEY` directly.

HarlemSquirrel committed:
Update Vivotek camera component

- Build camera and stream source ouside VivotekCam
- Remove unnecessary _stream_source attribute

MartinHjelmare commented on homeassistant/components/vivotek/camera.py:
```diff
@@ -0,0 +1,123 @@
+        self._stream_source = stream_source
+
+        self._brand = "Vivotek"
+        self._supported_features = SUPPORT_STREAM
```
We don't need instance attributes for brand and supported features. Just return those constants directly in the respective properties.

MartinHjelmare commented:
Please solve the merge conflict.

HarlemSquirrel committed:
Update Vivotek camera component

- Move brand to constant
- Move _supported_features to property

HarlemSquirrel committed:
Merge remote-tracking branch 'upstream/dev' into camera-vivotek

HarlemSquirrel committed:
Update Vivotek camera compontent to remove unused property

MartinHjelmare reviewed (approved):
Great!
